File: web/scripts/rsi_dispositions_test.js
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

// Hermetic tests for the 🎯T354 RHS coach list model.
// Run: node web/scripts/rsi_dispositions_test.js
//
// Oracle for the honesty criteria: an empty store renders a calm empty
// state (not an error), a fetch failure renders an error (not "nothing
// found"), and pending / ignored / filed rows each carry the fields the
// owner needs to see what the coach judged and what was done with it.

'use strict';

const assert = require('assert');
const RD = require('./rsi_dispositions.js');
const FT = require('./frontier_table.js');

let failed = 0;
function test(name, fn) {
  try {
    fn();
    console.log('ok  -', name);
  } catch (e) {
    failed++;
    console.error('FAIL-', name);
    console.error('    ', e && e.stack ? e.stack.split('\n').slice(0, 4).join('\n     ') : e);
  }
}

// Mirrors the shape GET /api/rsi/dispositions returns (server test covers
// that the Go side actually produces it).
const PAYLOAD = {
  total: 3,
  count: 3,
  pending: 1,
  path: '/tmp/state/rsi/dispositions.json',
  judgments: [
    {
      fingerprint: 'fp-pending',
      name: 'chat gap',
      observation: 'owner asked twice before a reply landed',
      severity: 'medium',
      delivered_at: '2026-08-09T03:00:00Z',
      disposition: 'pending',
      evidence: 'owner_chat:chatlog-2026-08-09 (chat_gap)',
    },
    {
      fingerprint: 'fp-ignored',
      name: 'phrase friction',
      observation: 'repeat phrasing in overseer replies',
      severity: 'low',
      delivered_at: '2026-08-09T02:00:00Z',
      disposition: 'ignore_with_reason',
      disposition_at: '2026-08-09T03:01:00Z',
      reason: 'one-off, no standing pattern',
    },
    {
      fingerprint: 'fp-filed',
      name: 'repair churn',
      observation: 'three follow-up commits on the same file',
      severity: 'high',
      mode: 'retro',
      delivered_at: '2026-08-09T01:00:00Z',
      disposition: 'file',
      disposition_at: '2026-08-09T03:02:00Z',
      target_id: 'T999',
    },
  ],
};

function rowsByFP(model) {
  const m = {};
  model.rows.forEach(function (r) { m[r.fingerprint] = r; });
  return m;
}

test('empty store is calm and empty, never an error', function () {
  const m = RD.normalizePayload({ judgments: [], total: 0, pending: 0, count: 0 });
  assert.strictEqual(m.empty, true);
  assert.strictEqual(m.error, '');
  assert.strictEqual(m.total, 0);
  assert.strictEqual(m.rows.length, 0);
  assert.ok(RD.emptyText().length > 0);
  // No counts line when there is nothing to count.
  assert.strictEqual(RD.countsText(m), '');
});

test('missing judgments array degrades to empty, not a crash', function () {
  const m = RD.normalizePayload(null);
  assert.strictEqual(m.empty, true);
  assert.strictEqual(m.error, '');
  assert.strictEqual(m.rows.length, 0);
});

test('fetch failure is a loud error, not "nothing found"', function () {
  const m = RD.normalizePayload(null, new Error('HTTP 500'));
  assert.strictEqual(m.error, 'HTTP 500');
  // The distinction the owner depends on: a failure must not read as empty.
  assert.strictEqual(m.empty, false);
  assert.strictEqual(m.rows.length, 0);
});

test('pending / ignored / filed shapes all reach the owner list', function () {
  const m = RD.normalizePayload(PAYLOAD);
  assert.strictEqual(m.total, 3);
  assert.strictEqual(m.pending, 1);
  assert.strictEqual(m.empty, false);
  assert.strictEqual(m.path, '/tmp/state/rsi/dispositions.json');
  const by = rowsByFP(m);

  const pending = by['fp-pending'];
  assert.strictEqual(pending.disposition, 'pending');
  assert.strictEqual(pending.dispositionLabel, 'pending');
  assert.strictEqual(pending.pending, true);
  assert.strictEqual(pending.severity, 'medium');
  assert.strictEqual(pending.title, 'chat gap');
  assert.ok(pending.deliveredAt > 0);
  assert.ok(pending.evidence.indexOf('chatlog-2026-08-09') >= 0);
  assert.strictEqual(RD.detailText(pending), '');

  const ignored = by['fp-ignored'];
  assert.strictEqual(ignored.dispositionLabel, 'ignored');
  assert.strictEqual(ignored.pending, false);
  // An ignore with no visible reason is indistinguishable from a drop.
  assert.ok(RD.detailText(ignored).indexOf('one-off') >= 0);

  const filed = by['fp-filed'];
  assert.strictEqual(filed.dispositionLabel, 'filed');
  assert.strictEqual(filed.targetID, 'T999');
  assert.strictEqual(filed.retro, true, 'T353 retro provenance must survive');
  assert.ok(RD.detailText(filed).indexOf('🎯T999') >= 0);

  assert.strictEqual(RD.countsText(m), '3 judgments · 1 pending');
});

test('rows sort newest delivery first, then severity', function () {
  const m = RD.normalizePayload(PAYLOAD);
  assert.deepStrictEqual(
    m.rows.map(function (r) { return r.fingerprint; }),
    ['fp-pending', 'fp-ignored', 'fp-filed']);

  // Same delivery instant → higher severity first, then fingerprint.
  const tied = RD.normalizePayload({
    judgments: [
      { fingerprint: 'b', severity: 'low', delivered_at: '2026-08-09T01:00:00Z' },
      { fingerprint: 'a', severity: 'high', delivered_at: '2026-08-09T01:00:00Z' },
    ],
  });
  assert.deepStrictEqual(
    tied.rows.map(function (r) { return r.fingerprint; }), ['a', 'b']);
});

test('blank / unknown disposition reads as pending', function () {
  const m = RD.normalizePayload({
    judgments: [
      { fingerprint: 'blank', delivered_at: '2026-08-09T01:00:00Z' },
      { fingerprint: 'bogus', disposition: 'nonsense', delivered_at: '2026-08-09T01:00:00Z' },
    ],
  });
  m.rows.forEach(function (r) {
    assert.strictEqual(r.disposition, 'pending', r.fingerprint);
    assert.strictEqual(r.pending, true, r.fingerprint);
  });
  // Derived when the server omits the count.
  assert.strictEqual(m.pending, 2);
});

test('Go zero time is absent, not 0001', function () {
  assert.strictEqual(RD.parseTime('0001-01-01T00:00:00Z'), 0);
  assert.strictEqual(RD.parseTime(''), 0);
  assert.strictEqual(RD.parseTime(undefined), 0);
  assert.ok(RD.parseTime('2026-08-09T03:00:00Z') > 0);
});

test('a judgment with no name still renders a title', function () {
  const r = RD.row({ fingerprint: 'x', observation: 'only an observation' });
  assert.strictEqual(r.title, 'only an observation');
  const bare = RD.row({ fingerprint: 'y' });
  assert.ok(bare.title.length > 0, 'never an anonymous blank row');
});

test('coach is a real RHS tab alongside frontier and transcript', function () {
  assert.strictEqual(FT.nextBottomTab('frontier', 'coach'), 'coach');
  assert.strictEqual(FT.nextBottomTab('coach', 'frontier'), 'frontier');
  // Sticky on unknown clicks, and never silently swallows the coach pane.
  assert.strictEqual(FT.nextBottomTab('coach', 'bogus'), 'coach');
  assert.strictEqual(FT.nextBottomTab('transcript', 'bogus'), 'transcript');
  assert.strictEqual(FT.nextBottomTab('frontier', 'bogus'), 'frontier');
});

if (failed) {
  console.error(failed + ' failed');
  process.exit(1);
}
console.log('All rsi_dispositions tests passed');

File: internal/server/rsi_dispositions_test.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

package server

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/marcelocantos/jevons/internal/rsi"
)

func rsiDispositionsServer(t *testing.T) (*httptest.Server, string) {
	t.Helper()
	state := t.TempDir()
	s := New("test", state)
	mux := http.NewServeMux()
	s.RegisterRoutes(mux)
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)
	return srv, state
}

func getRSIJudgments(t *testing.T, url string) rsiJudgmentsResponse {
	t.Helper()
	resp, err := http.Get(url)
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("GET %s status %d", url, resp.StatusCode)
	}
	var out rsiJudgmentsResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		t.Fatal(err)
	}
	return out
}

// 🎯T354 oracle: an empty store is an honest empty list, not an error.
func TestRSIDispositionsEmptyStoreIsHonest(t *testing.T) {
	srv, _ := rsiDispositionsServer(t)
	got := getRSIJudgments(t, srv.URL+"/api/rsi/dispositions")
	if got.Count != 0 || got.Total != 0 || got.Pending != 0 {
		t.Fatalf("empty store not honest: %+v", got)
	}
	if got.Judgments == nil {
		t.Fatal("judgments must serialize as [] not null")
	}
	if len(got.Judgments) != 0 {
		t.Fatalf("judgments=%+v", got.Judgments)
	}
}

// 🎯T354 oracle: pending + ignored + filed shapes all reach the owner list with
// severity, delivered_at, fingerprint, evidence summary, reason and target_id.
func TestRSIDispositionsPendingIgnoredFiledShapes(t *testing.T) {
	srv, state := rsiDispositionsServer(t)
	store, err := rsi.OpenDispositionStore(state)
	if err != nil {
		t.Fatal(err)
	}
	base := time.Date(2026, 8, 9, 3, 0, 0, 0, time.UTC)
	err = store.RecordDelivered([]rsi.Judgment{
		{
			Fingerprint: "fp-pending",
			Name:        "chat gap",
			Observation: "owner asked twice before a reply landed",
			Severity:    "medium",
			Evidence: []rsi.EvidenceRef{
				{Source: "owner_chat", SourceID: "chatlog-2026-08-09", Kind: "chat_gap"},
			},
		},
		{
			Fingerprint: "fp-ignored",
			Name:        "phrase friction",
			Observation: "repeat phrasing in overseer replies",
			Severity:    "low",
		},
		{
			Fingerprint: "fp-filed",
			Name:        "repair churn",
			Observation: "three follow-up commits on the same file",
			Severity:    "high",
			Mode:        rsi.RetroModeLabel,
		},
	}, base)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := store.SetDisposition(rsi.SetDispositionArgs{
		Fingerprint: "fp-ignored",
		Disposition: rsi.DispositionIgnoreWithReason,
		Reason:      "one-off, no standing pattern",
		Now:         base.Add(time.Minute),
	}); err != nil {
		t.Fatal(err)
	}
	if _, err := store.SetDisposition(rsi.SetDispositionArgs{
		Fingerprint: "fp-filed",
		Disposition: rsi.DispositionFile,
		TargetID:    "T999",
		TargetCwd:   state,
		Now:         base.Add(2 * time.Minute),
	}); err != nil {
		t.Fatal(err)
	}

	got := getRSIJudgments(t, srv.URL+"/api/rsi/dispositions")
	if got.Total != 3 || got.Count != 3 {
		t.Fatalf("counts total=%d count=%d", got.Total, got.Count)
	}
	if got.Pending != 1 {
		t.Fatalf("pending=%d, want 1", got.Pending)
	}
	byFP := map[string]rsi.DispositionEntry{}
	for _, e := range got.Judgments {
		byFP[e.Fingerprint] = e
	}

	pending, ok := byFP["fp-pending"]
	if !ok {
		t.Fatal("pending judgment missing from owner list")
	}
	if pending.Disposition != rsi.DispositionPending {
		t.Fatalf("pending disposition=%q", pending.Disposition)
	}
	if pending.Severity != "medium" || pending.Observation == "" || pending.DeliveredAt.IsZero() {
		t.Fatalf("pending row incomplete: %+v", pending)
	}
	if pending.Evidence != "owner_chat:chatlog-2026-08-09 (chat_gap)" {
		t.Fatalf("evidence summary=%q", pending.Evidence)
	}

	ignored := byFP["fp-ignored"]
	if ignored.Disposition != rsi.DispositionIgnoreWithReason {
		t.Fatalf("ignored disposition=%q", ignored.Disposition)
	}
	if ignored.Reason != "one-off, no standing pattern" {
		t.Fatalf("ignored reason=%q", ignored.Reason)
	}

	filed := byFP["fp-filed"]
	if filed.Disposition != rsi.DispositionFile || filed.TargetID != "T999" {
		t.Fatalf("filed row=%+v", filed)
	}
	if filed.Mode != rsi.RetroModeLabel {
		t.Fatalf("retro provenance lost: mode=%q", filed.Mode)
	}

	// Filter narrows to one bucket without losing the global pending count.
	only := getRSIJudgments(t, srv.URL+"/api/rsi/dispositions?disposition=pending")
	if only.Count != 1 || only.Judgments[0].Fingerprint != "fp-pending" {
		t.Fatalf("filtered list=%+v", only)
	}
	if only.Total != 3 || only.Pending != 1 {
		t.Fatalf("filtered totals total=%d pending=%d", only.Total, only.Pending)
	}
}

// 🎯T354: an unknown filter is a 400, not a silently empty list (which would
// read as "the coach found nothing").
func TestRSIDispositionsRejectsUnknownFilter(t *testing.T) {
	srv, _ := rsiDispositionsServer(t)
	resp, err := http.Get(srv.URL + "/api/rsi/dispositions?disposition=bogus")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusBadRequest {
		t.Fatalf("status=%d, want 400", resp.StatusCode)
	}
}
